edb/common/signalctl.py

```diff
@@ -162,46 +173,42 @@ async def wait_for(self, fut, *, cancel_on=None):
-                if ex.args and isinstance(
-                    ex.args[0], (asyncio.CancelledError, SignalError)
-                ):
-                    break
-                ex = ex.__context__
+
+        except asyncio.CancelledError as ex:
+            # 2. "fut" is cancelled - this usually means we caught a signal,
+            #    but it could also be other reasons, see below.
+            if cancelled_by is not None:
+                # Event 2 happened at least once
+                if outer_cancelled_at_last:
+                    # If event 3 is the last event, the outer code is probably
+                    # expecting a CancelledError, e.g. asyncio.wait_for().
+                    # Therefore, we just raise it with signal errors attached.
+                    ex.__context__ = cancelled_by
+                    rv = ex
+                else:
+                    # If event 2 is the last event, simply raise the grouped
+                    # signal errors, attaching the CancelledError to reveal
+                    # where the signals hit the user code.
+                    cancelled_by.__cause__ = ex
+                    rv = cancelled_by
             else:
-                # 2. We didn't find any clue, this likely means the chain was
-                #    lost or broken in the Task code, let's just re-raise.
+                # Neither event 2 nor 3 happened, the user code cancelled
+                # itself, simply propagate the same error.
                 raise
 
-        # 3. The first CancelledError that carries event 2 or 3 is found, let's
-        #    construct the proper error chain here based on the carriers.
-        current = rv = ex.args[0].with_traceback(ex.__traceback__)
-        ctx = ex.__context__
-        while ctx is not None:
-            if ctx.args and isinstance(ctx.args[0], asyncio.CancelledError):
-                # Event 3: keep the traceback of the original cancellation
-                current.__context__ = ctx.args[0]
-            elif ctx.args and isinstance(ctx.args[0], SignalError):
-                # Event 2: as SignalError was never directly raised by now,
-                # it doesn't have a traceback yet. We're just borrowing the
-                # traceback from its carrier CancelledError which reveals the
-                # actual line of code where the signal was caught.
-                current.__context__ = ctx.args[0].with_traceback(
-                    ctx.__traceback__
-                )
+        except Exception as e:
+            # 3. For any other errors, we just raise it with the signal errors
+            #    attached as __context__ if event 2 happened.
+            if cancelled_by is not None:
+                e.__context__ = cancelled_by
+                rv = e
             else:
-                # Preserve whatever else is in the chain.
-                current.__context__ = ctx
-            current = current.__context__
-            ctx = ctx.__context__
+                raise
```

Comment:
Should it be bare `raise` or `raise e`? I suspect the former, in which case we can drop line 209

Reply:
It should be bare `raise` just because the traceback is one line shorter than `raise e` when `cancelled_by` is None.

If `cancelled_by` is not None, it is almost equivalent to:

```python
        except Exception as e:
            if cancelled_by is not None:
                e.__context__ = cancelled_by
            raise e  # looks like reraising itself won't reset __context__
```

Reply:
yeah, so drop line 209